keps/sig-network/3070-reserved-service-ip-range/README.md

```diff
@@ -0,0 +1,436 @@
+// randomScanReservedStrategy choose a random address from the provided big.Int and then scans
+// forward looking for the next available address. The big.Int range is subdivided so it will try
+// to allocate first from the reserved block of addresses (it will wrap the reserved subrange if necessary).
+// If there is no free address it will try to allocate one from the reserved block too.
+type randomScanReservedStrategy struct {
+	rand     *rand.Rand
+	reserved int
+}
+```
+
+### Test Plan
+
+This feature doesn't modify the cluster behavior, only the order on which dynamic IP are assigned to Services,
+there is no need for e2e or integration tests, unit tests are enough.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+
+#### Beta
+
+- Gather feedback from developers and community.
+- No issues reported.
+
+#### GA
+
+- 2 examples of real-world usage
+
+### Upgrade / Downgrade Strategy
+
+The feature only changes the allocation strategy, the bitmaps and the underlay logic remain the same,
+guaranteeing full compatibility and no risk rolling back the feature.
+
+### Version Skew Strategy
+
+Same as with upgrade, there is no change in the bitmaps, keeping the feature 100% compatible.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: ServiceIPStaticSubrange
+  - Components depending on the feature gate: kube-apiserver
+- [x] Other
+  - Describe the mechanism:
+  - The feature require to restart the apiserver to use the new allocation strategy for Services, however,
+  it will not require downtime in multi-control-plane environments.
+    
+###### Does enabling the feature change any default behavior?
+
+Dynamic allocated IPs will not be chosen from the first 32 values of the Service IP range, that is transparent to the user,
+since the IP is and was already random, it doesn't matter from within the range the IP is obtained.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Nothing, this only affects new requests and how these new request obtain the dynamically assigned IPs.
+
+###### Are there any tests for feature enablement/disablement?
+
+No need to add new test, current tests should keep working as today.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+Current Services will not be affected, it doesn't matter in which range their IPs are allocated.
+This only affects the creation of new Services without an IP specified, there is no risk of impact on running workloads.
+
+###### What specific metrics should inform a rollback?
+
+The allocation logic already has the following metrics:
```

Comment:
Do we want new metrics to indicate "forced to use reserved range" ?

Reply:
I've tried to add new metrics but its complicated, since this behavior is deep in the bitmap allocation and the metrics are one layer above on the ip allocator .... inheritance problems :shrug: 

I'll add a comment explaining it

Reply:
> I've tried to add new metrics but its complicated, since this behavior is deep in the bitmap allocation and the metrics are one layer above on the ip allocator .... inheritance problems

Hmm - can't we report them somewhat post-factum (when we're returning the actual result we already know what IP we allocated and if it was statically requested or not)?

Reply:
do you mean something like `randomly_allocated_ips` gauge and `randomly_allocation_total` counter ?
that is easy to do, will that work for you?

Reply:
> do you mean something like randomly_allocated_ips gauge and randomly_allocation_total counter ?
that is easy to do, will that work for you?

Can you clarify? I think these can be interpreted differently.

Reply:
we have these metrics now
```
  - allocated_ips
  - available_ips
  - allocation_total
  - allocation_errors_total
```
we can add two new ones
```
randomly_allocated_ips
randomly_allocation_total
```
`randomly_allocated_ips` reports the number of IPs in the allocator that are allocated "randomly" i.e. with ClusterIP empty, it reports current value.
`randomly_allocation_total` is a counter, every time that you allocate a ClusterIP randomly it increases, the user can play with the current metrics and these new ones to obtain the IPs allocated statically

Reply:
Thanks for clarifiction.

Yes - those metric will be very useful.

What I had on my mind (please let me know if this doesn't make any sense because of some reason) is actually an extension of that by:
- adding a label to that metric (like "from-static" or sth) to show if those random IPs were allocated from static range

Does that make sense?

Reply:
yeah, that is much simpler and easier, I will update the kep with that